Crypto/PublicKey+Apple.mm

```diff
@@ -262,58 +262,75 @@ virtual int _sign(int/*mbedtls_md_type_t*/ mbedDigestAlgorithm,
+                (id)kSecAttrKeyType:            (id)kSecAttrKeyTypeRSA,
+                (id)kSecAttrKeyClass:           (id)kSecAttrKeyClassPrivate,
+                (id)kSecAttrApplicationLabel:   publicKeyHash,
+                (id)kSecReturnRef:              @YES
+            });
+            if (!privateKeyRef) {
+                CFAutorelease(publicKeyRef);
+                return nullptr;
+            }
+            
+            auto keySize = unsigned(8 * SecKeyGetBlockSize(privateKeyRef));
+            return new KeychainKeyPair(keySize, publicKeyRef, privateKeyRef);
+        }
     }
 
 
     Retained<PersistentPrivateKey> PersistentPrivateKey::withPublicKey(PublicKey* publicKey) {
         @autoreleasepool {
-            // First look up a SecCertificateRef from the public-key digest. (We ought to be able
-            // to look up a SecIdentityRef directly, but using kSecClassIdentity doesn't work;
-            // the search matches all the identities in the Keychain...)
-            auto certRef = (SecCertificateRef)findInKeychain(@{
-                (id)kSecClass:              (id)kSecClassCertificate,
-                (id)kSecAttrPublicKeyHash:  publicKeyHash(publicKey),
-                (id)kSecReturnRef:          @YES,
+            // Lookup private key by using the public key hash:
+            auto privateKeyRef = (SecKeyRef)findInKeychain(@{
+                (id)kSecClass:                  (id)kSecClassKey,
+                (id)kSecAttrKeyType:            (id)kSecAttrKeyTypeRSA,
+                (id)kSecAttrKeyClass:           (id)kSecAttrKeyClassPrivate,
+                (id)kSecAttrApplicationLabel:   publicKeyHash(publicKey),
+                (id)kSecReturnRef:              @YES
             });
-            if (!certRef)
+            if (!privateKeyRef)
                 return nullptr;
-            CFAutorelease(certRef);
-
-            // Get the identity and then the private key:
-            SecIdentityRef identityRef;
-            checkOSStatus(SecIdentityCreateWithCertificate(nullptr, certRef, &identityRef),
-                          "SecIdentityCreateWithCertificate", "get private key from keychain");
-            CFAutorelease(identityRef);
-
-            SecKeyRef privateKeyRef;
-            checkOSStatus(SecIdentityCopyPrivateKey(identityRef, &privateKeyRef),
-                          "SecIdentityCopyPrivateKey", "get private key from keychain");
-
-            // Get the public key from the cert, not from the private key, because calling
-            // SecKeyCopyPublicKey results in a key ref that doesn't allow its data to be read,
-            // for some reason (bug?)
-
-            SecKeyRef publicKeyRef;
-            checkOSStatus(SecCertificateCopyPublicKey(certRef, &publicKeyRef),
-                          "SecCertificateCopyPublicKey", "get private key from keychain");
+            
+            auto publicKeyRef = SecKeyCopyPublicKey(privateKeyRef);
```

Comment:
same, any reason why not auto release?

Reply:
Because we want to release it here only when there is an error.

Reply:
Then why `CFAutoRelease` instead of `CFRelease` ?

Reply:
It was supposed to be CFRelease. It's a bug.